File: src/main/scala/demos/fp/order/hiding/Order.scala
package demos.fp.order.hiding

class Order(val lines: List[OrderLine] = List(), orderId: String):
    def add(productId: String, amount: Int): Order =
        val existingOrderLineAndIndex = this.lines.zipWithIndex.find(_._1.productId == productId)
        val updatedLines = existingOrderLineAndIndex match
            case None =>
                val newLine = OrderLine(productId, amount)
                this.lines.appended(newLine)
            case Some((existingOrderLine, index)) =>
                val updatedLine = existingOrderLine.add(amount)
                this.lines.updated(index, updatedLine)
        Order(lines = updatedLines, orderId = this.orderId)

    override def toString: String = s"Order(lines=${this.lines}, orderId=${this.orderId})"

File: src/main/java/demos/oop/order/hiding/Order.java
package demos.oop.order.hiding;

import java.util.ArrayList;
import java.util.List;

class Order {
    private List<OrderLine> lines;
    private String orderId;

    public Order(String orderId) {
        this.orderId = orderId;
        this.lines = new ArrayList<OrderLine>();
    }

    public void add(String productId, int amount) {
        OrderLine existingOrderLine = this.findExistingOrderLine(productId);
        if (existingOrderLine == null) {
            OrderLine newOrderLine = new OrderLine(productId, amount);
            newOrderLine.validate();
            this.lines.add(newOrderLine);
        } else {
            existingOrderLine.add(amount);
        }
    }

    private OrderLine findExistingOrderLine(String productId) {
        for (OrderLine orderLine : this.lines) {
            if (orderLine.getProductId().equals(productId)) {
                return orderLine;
            }
        }
        return null;
    }

    public List<OrderLine> getLines() {
        return List.copyOf(this.lines);
    }

    public String getOrderId() {
        return orderId;
    }
}
